contrib/babelfishpg_tsql/src/linked_servers.c

```diff
@@ -0,0 +1,1024 @@
+			 * directly be stored in a PG tuple store, so they do not need our TDS side receiver magic.
+			 */
+			PG_RETURN_TEXT_P(cstring_to_text_with_len((char *)val, len));
+			break;
+		case TSQL_DATETIME:
+		case TSQL_DATETIMN:
+			if (*pltsql_protocol_plugin_ptr && (*pltsql_protocol_plugin_ptr)->get_datum_from_byte_ptr)
+			{
+				StringInfo pbuf;
+
+				pbuf = palloc(sizeof(StringInfoData));
+				/*
+				 * Rather than copying data around, we just set up a phony
+				 * StringInfo pointing to the correct portion of the TDS message
+				 * buffer. 
+				 */
+				pbuf->data = (char *)val;
+				pbuf->maxlen = 8;
+				pbuf->len = 8;
+				pbuf->cursor = 0;
+
+				return (*pltsql_protocol_plugin_ptr)->get_datum_from_byte_ptr(pbuf, TSQL_DATETIMN, 0);
+			}
+			break;
+		case TSQL_SMALLDATETIME:
+			if (*pltsql_protocol_plugin_ptr && (*pltsql_protocol_plugin_ptr)->get_datum_from_byte_ptr)
+			{
+				StringInfo pbuf;
+
+				pbuf = palloc(sizeof(StringInfoData));
+				/*
+				 * Rather than copying data around, we just set up a phony
+				 * StringInfo pointing to the correct portion of the TDS message
+				 * buffer. 
+				 */
+				pbuf->data = (char *)val;
+				pbuf->maxlen = 4;
+				pbuf->len = 4;
+				pbuf->cursor = 0;
+
+				return (*pltsql_protocol_plugin_ptr)->get_datum_from_byte_ptr(pbuf, TSQL_SMALLDATETIME, 0);
+			}
+			break;
+		case TSQL_DATE:
+			if (*pltsql_protocol_plugin_ptr && (*pltsql_protocol_plugin_ptr)->get_datum_from_date_time_struct)
+			{
+				LS_TDS_DATETIMEALL date;
+
+				memcpy(&date, (LS_TDS_DATETIMEALL *) val, sizeof(LS_TDS_DATETIMEALL));
+
+				/* No optional attribute */
+				return (*pltsql_protocol_plugin_ptr)->get_datum_from_date_time_struct(0, date.date, TSQL_DATE, 0);
+			}
+			break;
+		case TSQL_TIME:
+			if (*pltsql_protocol_plugin_ptr && (*pltsql_protocol_plugin_ptr)->get_datum_from_date_time_struct)
+			{
+				LS_TDS_DATETIMEALL time;
+
+				memcpy(&time, (LS_TDS_DATETIMEALL *) val, sizeof(LS_TDS_DATETIMEALL));
+
+				/* optional attribute here is scale */
+				return (*pltsql_protocol_plugin_ptr)->get_datum_from_date_time_struct(time.time, 0, TSQL_TIME, 7);
+			}
+			break;
+		case TSQL_DECIMAL:
+		case TSQL_NUMERIC:
+			if (*pltsql_protocol_plugin_ptr && (*pltsql_protocol_plugin_ptr)->get_datum_from_byte_ptr)
+			{
+				LS_TDS_NUMERIC *numeric;
+				StringInfo pbuf;
+				int n, i = 0;
+				char numeric_bytes[16] = {0x00}; /* max storage bytes for numeric - sign byte = 17 - 1 = 16 */
+
+				numeric = (LS_TDS_NUMERIC *)val;
+
+				pbuf = palloc(sizeof(StringInfoData));
+				/*
+				 * Rather than copying data around, we just set up a phony
+				 * StringInfo pointing to the correct portion of the TDS message
+				 * buffer. 
+				 */
+				pbuf->data = (char *)(numeric->array);
+
+				n = tds_numeric_bytes_per_prec[numeric->precision] - 1;
+
+				/* reverse copy 'n' bytes after 1st byte (sign byte) */
+				for (i = 0; i < n; i++)
+					numeric_bytes[i] = numeric->array[n-i];
```

Comment:
this is not required, data already points to numeric array, just reverse in place

Reply:
doing reverse in place now